File: leetcode/796-rotate-string/main.go
package main

import (
	"fmt"
	"strings"
)

/*
	1st approach
	- 2A must contain a rotated string
	Time	O(n) indexOf
	Space	O(n)
	0ms beats 100%
*/
func rotateString(A string, B string) bool {
	C := A + A
	return strings.Index(C, B) > -1 && len(B) == len(A)
}

func main() {
	a := "abcdeabcde"
	fmt.Println(strings.Index(a, "abced"))
}

File: leetcode/796-rotate-string/main.py
"""
    1st approach:
    - rotate A to see if B appears

    Time    O(n)
    Space   O(1)
    12 ms, faster than 90.20%
"""


class Solution(object):
    def rotateString(self, A, B):
        """
        :type A: str
        :type B: str
        :rtype: bool
        """
        if len(A) != len(B):
            return False
        if len(A) == len(B) and len(A) == 0:
            return True
        i = 0
        while i < len(A):
            A = A[1:] + A[0]
            if A == B:
                return True
            i += 1
        return False


"""
    2nd approach:
    - 2A must contain a rotated string

    Time    O(n)
    Space   O(1)
    12 ms, faster than 90.20%
"""


class Solution(object):
    def rotateString(self, A, B):
        """
        :type A: str
        :type B: str
        :rtype: bool
        """
        C = A + A
        return len(A) == len(B) and B in C
